### DISCUSSION TITLE
@stream initialCount validation

### DISCUSSION THREAD

**Discussion by robrichard** [graphql/defer-stream-wg#24] [Architectural Decision Records]
## Context
From @benjie https://github.com/graphql/graphql-spec/pull/742#pullrequestreview-772422931
> There doesn't seem to be any validation rule enforcing that initialCount (if present) is greater than or equal to 0. I suggest adding this validation.

## Decision
There was agreement in the Dec 2021 WG meeting to throw a field error when a negative initialCount is encountered during execution. 

## Status
- [x] defined in spec
- [x] implemented in graphql-js

**Comment by robrichard** [comment]
@benjie I think this validation would require variables to be known? Does that mean it must be done during execution and not during validation? Are there examples of anything else like this?

For example, this document is valid

```graphql
query ($myInitialCount: Int) {
  listField @stream(initialCount: $myInitialCount)
}
```
But not if sent with the arguments `{"myInitialCount": -1}`

**Comment by benjie** [comment]
Yeah, we can’t do it at the validation stage. Maybe just add text stating that negative numbers are treated as zero. Or even that the behaviour of negative numbers is undefined (which allows for them to do something specific in future).

**Reply by robrichard** [reply]
I think it makes sense to go with negative numbers being treated as zero. The logic in the spec is only comparing against `initialCount` using "is greater or equal to" so that's effectively how they are already treated. I updated the spec PR to explicitly note this and added a test in graphql-js

**Reply by maraisr** [reply]
If a value is in fact passed in, can it not fail validation of less than zero?

**Reply by benjie** [reply]
That would be inconsistent - passing a variable and an explicit argument  with the same value (assuming typed correctly) should have the same effect.

**Reply by michaelstaib** [reply]
It still feels fishy I have to say. What about a runtime error if the value is incorrect?

**Reply by maraisr** [reply]
I think there's 2 types of validation, query validation and input/runtime valuation. 

An "incorrect" input would still pass query validation, but a server could decide to fail input validation, but that is "consumer" side from the eyes of graphql.

**Reply by benjie** [reply]
Agree. This problem isn't really unique to `@stream`; e.g. Relay connections with `first: -1` don't make sense. The difference being that the argument is to a directive versus a field. Quite a few GraphQL services have a second validation phase that runs "per-request" rather than "per-document" - this is required to do things like query cost limits, and is also useful for pagination limits, etc. The Guild in particular are keen to make this second phase more official. Relevant: https://github.com/graphql/graphql-js/pull/3370#issuecomment-972925813

**Comment by robrichard** [comment]
There was agreement in the [Dec 2021 WG meeting](https://github.com/graphql/graphql-wg/blob/main/notes/2021-12-02.md#deferstream-updatesissuesdiscussions-30m-rob-richard) to throw a field error when a negative initialCount is encountered during execution.

**Reply by maraisr** [reply]
Not related; super stoked to have you at the helm of these things @robrichard ❤️
